src/irmin-pack/pack_value_intf.ml

```diff
@@ -56,17 +76,21 @@ module type Sigs = sig
     val of_magic_exn : char -> t
     val version : t -> Version.t
     val pp : t Fmt.t
+
+    val length_header_exn : t -> length_header
+    (** Raises an exception on [Contents], as the availability of a length
+        header is user defined. *)
```

Comment:
Isn't length_header (... `Sometimes of [kind -> ...]) something that already takes a kind? But this function also takes another a kind argument? (Are these different kinds?) 

On contents ... "availability of a length header is user defined". Does this mean that some contents do have a length header, and some don't? Or that they either all do or all don't, and at this point in the code we are not sure?

Reply:
The situation is made more complex by having different instantiations of `Pack_store.S` (with different `Pack_value.S`), each of which may be responsible for one or more different `Pack_value.Kind.t`s. We have:
- the contents store (built [here](https://github.com/CraigFe/irmin/blob/dev/src/irmin-pack/ext.ml#L47)), responsible for the user's contents type (of kind `Contents`).
- the node store (built [here](https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin-pack/inode.ml#L1783)), responsible for the various types of node object (`Inode_...`), some of which have length headers and some of which don't.
- the commit store (built [here](https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin-pack/ext.ml#L79)), which looks after `Commit_v0` (with no length header), and `Commit_v1` (with a length header).

> On contents ... "availability of a length header is user defined". Does this mean that some contents do have a length header, and some don't? Or that they either all do or all don't, and at this point in the code we are not sure?

Whether or not the contents type has a length header is dependent on the configuration passed to the store implementation ([here](https://github.com/mirage/irmin/blob/61827d70978ce46e9df3576eb6272a1e0c7e58a3/src/irmin-pack/conf.mli#L23-L30)), but there is only one contents type for any given store. This means that, at functor instantiation time, we may be guaranteed that the contents value is headerless (without needing to inspect the `kind`, which is guaranteed by construction to be `Contents`), which is where the `` `Never`` case comes from [here](https://github.com/mirage/irmin/blob/craig%40pack-direct-pointers/src/irmin-pack/pack_value.ml#L108-L111).

We don't actually use this `` `Never `` guarantee for much at the moment because of the way we eagerly read child entry prefixes in the pack store, so it's potentially worth removing.

Reply:
> We don't actually use this `Never guarantee for much at the moment because of the way we eagerly read child entry prefixes in the pack store, so it's potentially worth removing.

Now done in this PR.